app/Model/Assurance.php

```diff
@@ -0,0 +1,478 @@
+      $assurance_found = array_filter(
+        $env_attributes,
+        static function($value) {
+          return ($value['attribute'] === 'assurances:value');
+        }
+      );
+      if (!empty($assurance_found)) {
+        // XXX Pop the intermediate index level
+        $assurance_found = array_pop($assurance_found);
+        $assurance_env_map = !empty($assurance_found['env_name']) ? $assurance_found['env_name'] : "";
+      }
+    }
+
+    return $assurance_env_map;
+  }
+
+  /**
+   * Check if the environnmental attribute has value. If this is the case check if it is single valued or multi valued
+   *
+   * @param string $env_value
+   * @return bool|null  true for MULTI valued | false for SINGLE valued | null for NO value
+   *
+   * @depends Shibboleth SP configuration. Currently we assume that the delimiter is the default one `;`. The semicolon.
+   */
+  public function isEnvMultiVal($env_value) {
+    if(!empty($env_value)) {
+      $env_value_vals = explode(";", $env_value);
+      return (count($env_value_vals) > 1) ? true : false;
+    }
+    return null;
+  }
+
+  /**
+   * Get Assurance Environment Value
+   *
+   * @return string
+   */
+  public function getEnvValues() {
+    // Get Assurance mappings
+    $assurance_value = "";
+    $assurance_map = $this->getEnvAssuranceMapping();
+    // Get Subject DN env value if available
+    if(!empty($assurance_map)) {
+      $assurance_value = getenv($assurance_map);
+      $assurance_value = !empty($assurance_value) ? $assurance_value : "";
+    }
+
+    return $assurance_value;
+  }
+
+  /**
+   * @param string $identifier      OrgIdentity Identifier constructed from the IdP
+   */
+  public function syncByIdentifier($identifier) {
+    $current_assurances_list = array();
+    $current_assurances = $this->getAssurancesByOrgIdentityIdentifier($identifier);
+    $active_login_orgs = $this->getOrgIdentityByIdentifier($identifier);
+    $assurance_env_value = $this->getEnvValues();
+    // XXX Is the assurance attribute multi valued?
+    $is_assurance_val_multi_val = $this->isEnvMultiVal($assurance_env_value);
+
+    // XXX New Assurance list is empty
+    if(is_null($is_assurance_val_multi_val)) {
+      foreach($current_assurances as $assurance) {
+        // Soft Delete everything
+        // Since i am using delete function, Changelog behaviour should work
+        $this->delete($assurance['Assurance']['id']);
+      }
+      // Cleared everything, return
+      return;
+    }
+
+    // XXX The Assurance is MULTI valued
+    $new_assur_values = array();
+    if($is_assurance_val_multi_val) {
+      $new_assur_values = explode(';', $assurance_env_value);
+    } else {
+      $new_assur_values[] = $assurance_env_value;
+    }
+    $this->log(__METHOD__ . "::Session(New) Assurance values => " . print_r($new_assur_values, true), LOG_DEBUG);
+
+
+    // XXX Filter out the entries that do not pass the REFEDS assertion criteria for IAP
+//    $assurance_iap = array_filter(
```

Comment:
Do we need these comments?

Reply:
Yes. This is the validation of REFEDS assurance profiles. These will be enabled in the future

Reply:
οκ